blocks/library/categories/index.js

```diff
@@ -0,0 +1,241 @@
+/**
+ * WordPress dependencies
+ */
+import { Component } from 'element';
+import { Placeholder, Spinner } from 'components';
+import { __ } from 'i18n';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+import './block.scss';
+import { registerBlockType } from '../../api';
+import { getCategories } from './data.js';
+import InspectorControls from '../../inspector-controls';
+import ToggleControl from '../../inspector-controls/toggle-control';
+import BlockDescription from '../../block-description';
+import BlockControls from '../../block-controls';
+import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+
+registerBlockType( 'core/categories', {
+	title: __( 'Categories' ),
+
+	icon: 'list-view',
+
+	category: 'widgets',
+
+	defaultAttributes: {
+		showPostCounts: false,
+		displayAsDropdown: false,
+		showHierarchy: false,
+	},
+
+	getEditWrapperProps( attributes ) {
+		const { align } = attributes;
+		if ( 'left' === align || 'right' === align || 'full' === align ) {
+			return { 'data-align': align };
+		}
+	},
+
+	edit: class extends Component {
+		constructor() {
+			super( ...arguments );
+
+			this.state = {
+				categories: [],
+			};
+
+			this.categoriesRequest = getCategories();
+
+			this.categoriesRequest
+				.then( categories => this.setState( { categories } ) );
+
+			this.toggleDisplayAsDropdown = this.toggleDisplayAsDropdown.bind( this );
+			this.toggleShowPostCounts = this.toggleShowPostCounts.bind( this );
+			this.toggleShowHierarchy = this.toggleShowHierarchy.bind( this );
+		}
+
+		toggleDisplayAsDropdown() {
+			const { displayAsDropdown } = this.props.attributes;
+			const { setAttributes } = this.props;
+
+			setAttributes( { displayAsDropdown: ! displayAsDropdown } );
+		}
+
+		toggleShowPostCounts() {
+			const { showPostCounts } = this.props.attributes;
+			const { setAttributes } = this.props;
+
+			setAttributes( { showPostCounts: ! showPostCounts } );
+		}
+
+		toggleShowHierarchy() {
+			const { showHierarchy } = this.props.attributes;
+			const { setAttributes } = this.props;
+
+			setAttributes( { showHierarchy: ! showHierarchy } );
+		}
+
+		getCategories( parentId = null ) {
+			const { categories } = this.state;
+			if ( ! categories.length ) {
+				return [];
```

Comment:
Yeah, good call about returning categories here - updated in 0fa3f2936c3a66442de9f8019659e713e416efe8. I'm more inclined to keep the condition though - no need to continue if categories are not loaded yet.